### ISSUE TITLE
HelloPangolinOffscreen does not create a propber surface

### ISSUE THREAD

**Issue by christian-rauch** [stevenlovegrove/Pangolin#672]
On version `v0.6` the `HelloPangolinOffscreen` example produces a `window.png` with correct content (a cube) on Intel and Nvidia systems.

After the component refactoring (8cd9ad0b01b7a1aeb09e8497d7445cdb1bdc8214) and with the fix in 73b9586d290a6f81117af66ce69ca070ae522de5, this does not produce a valid surface to render anymore:
```
libpng warning: Image width is zero in IHDR
libpng warning: Image height is zero in IHDR
libpng error: Invalid IHDR data
PNG Error: Error during png creation.
```
and the resulting `window.png` is indeed empty.

**Comment by stevenlovegrove** [comment]
Hmm - I'm not even getting this far.
Mine is segfaulting in eglGetDisplay(), though that is perhaps because I'm running in a VM...

**Comment by christian-rauch** [comment]
Yes, that can be. I assume your VM does not have GPU acceleration enabled?

Does the simpler https://github.com/svenpilz/egl_offscreen_opengl work for you? On a newer Ubuntu you may need my PR from https://github.com/christian-rauch/egl_offscreen_opengl/tree/fix_make to build.

**Comment by christian-rauch** [comment]
Quickly tested this in a Ubuntu 20.04 VirtualBox VM with and without hardware 3D acceleration and I got the same results. Without hardware acceleration, `glxinfo | grep "OpenGL renderer string:"` should print something with `llvmpipe` to indicate that software rendering is used.

**Comment by stevenlovegrove** [comment]
I think mine is hardware accelerated, but it isn't a 'standard driver'
```
~/code/Pangolin$ glxinfo | grep "OpenGL renderer string:"
OpenGL renderer string: Parallels using AMD Radeon Pro Vega 56 OpenGL Engine (Compat)
```

To debug, I used my native coca windowing and made a couple of small tweaks to fix a small OSX issue around scaling 47e5f829d0a8f31072c2fd950734dbebd85c1a32 such that it now works there. I can't see why the EGL path is any different to before.

**Comment by christian-rauch** [comment]
I overlooked that the example actually demonstrates two ways to save the buffer. One way via `SaveWindowOnRender` stores the entire window on `FinishFrame`, and another way via `GlTexture::Save` stores a specific buffer on request.

After your patch, `color_buffer.Save("fbo.png", false);` correctly generates a `fbo.png`. After I add the `pangolin::SaveWindowOnRender("window");` back again, I get the same error message and result for `window.png`.

If you but `pangolin::SaveWindowOnRender("window");` back into the example, do you get a valid `window.png`?

**Comment by stevenlovegrove** [comment]
SaveWindowOnRender() doesn't make much sense in that example because it is rendering to a framebuffer, not the window buffer.

Looking again, I should actually remove the FinishFrame() because that is swapping the window buffer that we haven't even written to. If I put the SaveWindowOnRender back in anyway, it does save an image of the cube (though the scaling is wrong because it is using the dimensions of the scaled window buffer on OSX, not the correctly sized framebuffer dimensions). Getting something here is actually a bug - it is reading from the currently active PBO rather than what should have been read (the window buffer).

If I take out the FinishFrame() as I should, I get nothing because the save method on the window framebuffer is never called.

We might want two examples instead of one
a) Using an FBO object in a method outside of an ordinary main loop (regardless of EGL)
b) Using EGL to create an image without even creating a window using the instantiated EGL buffer itself rather than a new FBO

What do you think?

**Comment by stevenlovegrove** [comment]
Does your vector egl_configs have size > 0 (display_headless.cpp:81)? We should assert that count > 0 before accessing egl_configs[0].
